geode-core/src/test/java/org/apache/geode/internal/cache/wan/serial/SerialGatewaySenderQueueJUnitTest.java

```diff
@@ -0,0 +1,294 @@
+
+    TestableSerialGatewaySenderQueue queue = spy(realQueue);
+    queue.setGroupTransactionEvents(true);
+
+    doAnswer(invocation -> eventPair1)
+        .doAnswer(invocation -> eventPair2)
+        .doAnswer(invocation -> eventPair3)
+        .doAnswer(invocation -> null)
+        .when(queue).peekAhead();
+
+    doAnswer(invocation -> new SerialGatewaySenderQueue.EventsAndLastKey(
+        Arrays.asList(new Object[] {event4}), 2L))
+            .when(queue).getElementsMatching(any(), any(), anyLong());
+
+    List peeked = queue.peek(-1, 1);
+    assertEquals(4, peeked.size());
+  }
+
+  @Test
+  public void peekedMaxEventsWhenNotIsGroupTransactionEvents()
+      throws Exception {
+    GatewaySenderEventImpl event1 = createMockGatewaySenderEventImpl(1, false, region);
+    GatewaySenderEventImpl event2 = createMockGatewaySenderEventImpl(1, false, region);
+    GatewaySenderEventImpl event3 = createMockGatewaySenderEventImpl(2, false, region);
+    GatewaySenderEventImpl event4 = createMockGatewaySenderEventImpl(1, true, region);
+    GatewaySenderEventImpl event5 = createMockGatewaySenderEventImpl(2, true, region);
+
+    TestableSerialGatewaySenderQueue queue = new TestableSerialGatewaySenderQueue(sender,
+        TEST_REGION, metaRegionFactory);
+
+    queue.put(event1);
+    queue.put(event2);
+    queue.put(event3);
+    queue.put(event4);
+    queue.put(event5);
+
+    List peeked = queue.peek(3, 1000);
+    assertEquals(3, peeked.size());
+    List peekedAfter = queue.peek(3, 1000);
+    assertEquals(2, peekedAfter.size());
+  }
+
+  private static GatewaySenderEventImpl createMockGatewaySenderEventImpl(int transactionId,
+      boolean isLastEventInTransaction, Region region) {
+    GatewaySenderEventImpl event = mock(GatewaySenderEventImpl.class);
+    when(event.getTransactionId()).thenReturn(new TXId(null, transactionId));
+    when(event.makeHeapCopyIfOffHeap()).thenReturn(event);
+    when(event.isLastEventInTransaction()).thenReturn(isLastEventInTransaction);
+    when(event.getRegion()).thenReturn(region);
+    return event;
+  }
+
+  private Region createDistributedRegion(String regionName, Cache cache) {
+    AttributesFactory factory = new AttributesFactory();
+    factory.setScope(Scope.DISTRIBUTED_ACK);
+    factory.setDataPolicy(DataPolicy.NORMAL);
+    RegionAttributes attrs = factory.create();
+    InternalRegionArguments internalRegionArgs = new InternalRegionArguments();
+    return new TestLocalRegion(regionName, attrs, null, (InternalCache) cache, internalRegionArgs,
+        new TestStatisticsClock());
+  }
+
+  public class TestLocalRegion extends LocalRegion {
+    Map map = new ConcurrentHashMap();
+
+    public TestLocalRegion(String regionName, RegionAttributes attrs,
+        LocalRegion parentRegion, InternalCache cache,
+        InternalRegionArguments internalRegionArgs,
+        StatisticsClock statisticsClock) {
+      super(regionName, attrs, parentRegion, cache, internalRegionArgs, statisticsClock);
```

Comment:
I did not find a way to test `SerialGatewaySenderQueue` without creating a `LocalRegion`. The thing is the queue is backed up by a Region while in the ParallelGatewaySenderQueue it is an easily mockable class.

I could try mocking the LocalRegion but not sure how complex the mock would be.
I will give it a try.

I checked to see what an alternative integration test would look like but it looked to me it would need a lot of code too and did not find any example to have a clear idea on how to start.

I think it is best to try to have this as a unit test. Plus, I think it is cool that we also have the JUnit version for the SerialGatewaySenderQueue just as we do with the ParallelGatewaySenderQueue.

Reply:
Okay, if it's too much work to mock out the `LocalRegion` then this test can stay as it is. There are a lot of classes in Geode which are hard to unit test, and this PR isn't about trying to refactor them.

Reply:
I managed to mock the LocalRegion without much work. I will push a new commit shortly.

Reply:
Perfect! Thanks for taking the time to do this.